### ISSUE TITLE
request.form empty for certain form POST request payloads

### ISSUE THREAD

**Issue by dividebysandwich** [pallets/werkzeug#2734]
I have a weird one, and I admit I am not entirely sure if it's a problem in Werkzeug or maybe somewhere deeper:

I have an ecowitt weather station and am happily getting data pushed to a python script that works great. Recently someone else wanted to do the same, installed my script but... it just didn't work. We found that the only difference was that I was still on python 3.6.9 and Werkzeug 2.0.3, whereas he was on Python 3.9.2 and Werkzeug 2.3.6

What happens is that in the route handler, request.form is empty depending on how the form request looks in detail.

Here's the code snippet:

```
@app.route('/weather/report', methods=['POST'])
def receiveEcoWitt():
    print("Data size: " + str(len(request.form)))
    data = request.form
    for key in data:
        print(f"{key}: {data[key]}\n")
```

So far so easy, right? Now sending a POST request using a simple html form works fine as long as it looks like this:

```
POST /weather/report HTTP/1.1
Host: 192.168.178.208:8100
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/114.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Content-Type: application/x-www-form-urlencoded
Content-Length: 20
Origin: null
Connection: keep-alive
Upgrade-Insecure-Requests: 1

fname=John&lname=Doe

```
With that request, request.form is filled and the data can be parsed normally.
However, the Ecowitt station sends requests that have a different set of header fields, which looks like this:

```
POST /weather/report HTTP/1.1
HOST: 192.168.178.208
Connection: Close
Content-Type: application/x-www-form-urlencoded
Content-Length:502

PASSKEY=78D8DCE861B2164BAD38A473167ACFD2&stationtype=GW2000A_V2.2.4&runtime=1247&dateutc=2023-06-17+10:34:52&tempinf=87.98&humidityin=43&baromrelin=29.158&baromabsin=29.158&tempf=71.60&humidity=57&winddir=44&windspeedmph=0.00&windgustmph=0.00&maxdailygust=1.12&solarradiation=5.52&uv=0&rrain_piezo=0.000&erain_piezo=0.000&hrain_piezo=0.000&drain_piezo=0.000&wrain_piezo=0.000&mrain_piezo=0.000&yrain_piezo=0.000&ws90cap_volt=2.2&ws90_ver=132&wh26batt=0&wh90batt=2.98&freq=868M&model=GW2000A&interval=16

```

Sending that request results in an empty request.form. It's also no use to try and use request.get_data() or stuff like that. The method in both requests is POST.

What should happen: Both requests should work and request.form should be filled in both cases.

This works fine in Python 3.6.9 and Werkzeug 2.0.3

Environment:

Python v3.9.2
Werkzeug v2.3.6

**Comment by dividebysandwich** [comment]
Update: I recently updated the already-running weather station and now it has the same problem there with Werkzeug v2.3.6.
I don't know exactly which headers Ecowitt used to send back then, but there's clearly been a change in their headers in the latest firmware. Still, I don't think the form data should be empty in such cases, especially considering the request is otherwise accepted just fine.

**Comment by davidism** [comment]
In your second request example, you have `Content-Length:502` without a space, is that a typo? I don't think it should matter (and Werkzeug isn't the thing parsing the headers at that point), but it's the only thing that looks weird.

I can't reproduce this issue, although I don't have a way to reproduce that typo. `request.form` is populated correctly.

**Comment by djjudas21** [comment]
I have the same issue (which I logged here https://github.com/pallets/flask/issues/5168 and also discussed with another weather station hacker here https://github.com/bentasker/Ecowitt_to_InfluxDB/issues/1).

I'm on Python v3.11.3, Flask v2.3.2 and Werkzeug v2.3.6.

It seems we can deduce that something has changed with the format that the Ecowitt is sending in, but it also looks like this should be parsed correctly by Flask/Werkzeug

I'm running an Ecowitt WS2910 with firmware version v5.1.1, and the top item in the firmware changelog is "Optimize the http server" :man_facepalming:

**Comment by djjudas21** [comment]
Just to add, I've worked around this issue by placing an NGINX reverse proxy in front of Flask, with no special config, and it seems to magically fix the dodgy header in flight.

Both NGINX and the http echo server I used for testing were able to accept and correct the header, so it feels like Flask should be able to, too.

**Comment by davidism** [comment]
You still haven't provided a way to reproduce the issue locally, so I still can't do anything. Saying you "feel like Flask should be able" to handle it isn't enough.

It sounds like you might be running the development server, and assuming that is "Flask". This would only be an issue with Werkzeug if it also happens when running a production server such as Gunicorn, Waitress, or uWSGI. And you'd still need to provide a reproducible example.

The development server is a thin wrapper around Python's built-in `http.server`, issues with how it parses headers should be reported (again with a minimal reproducible example) to https://github.com/python/cpython.

**Comment by bentasker** [comment]
I'll look at raising a PR for the first in a bit.

@davidism let me know if you'd be willing to accept a PR for the second

**Comment by ThiefMaster** [comment]
Considering this regex is meant to remove stuff that's allowed in `int()` but not in a plain number, I think allowing surrounding whitespace makes perfect sense

**Comment by davidism** [comment]
PR to fix that is fine, probably just do `value = value.strip()` before the regex check.

**Comment by bentasker** [comment]
PR posted here: https://github.com/pallets/werkzeug/pull/2738

I'll look at getting one into Python for the root cause soon (or possibly tomorrow)

**Comment by pgjones** [comment]
Closed by #2764
